### PR TITLE
Deprecate `get_best_method` and `best_method_str` static methods in `QNode`

### PR DESCRIPTION
**Pull request by andrijapau**
**Context:**

#6399 added helper functions `get_best_diff_method` to `qml.workflow`. This means we are free to deprecate the old static methods out of `QNode`.

**Description of the Change:** 

Deprecation of the methods.

Tweaked the new `_get_gradient_fn` a bit.

**Benefits:** Cleaning up `QNode` and making it more user-friendly.

**Possible Drawbacks:** None

[sc-76083]

### CODE REVIEW

File: pennylane/workflow/qnode.py [hunk 1]
```diff
@@ -793,7 +826,12 @@ def get_best_method(
     @staticmethod
     @debug_logger
     def best_method_str(device: SupportedDeviceAPIs, interface: SupportedInterfaceUserInput) -> str:
-        """Similar to :meth:`~.get_best_method`, except return the
+        """
+        .. warning::
+
+            This method is deprecated in v0.40 and will be removed in v0.41.
```
Comment (mudit2812) [inline pennylane/workflow/qnode.py:832]: Same suggestion for the other deprecated methods as well.

```suggestion
            ``QNode.best_method_str`` is deprecated in v0.40 and will be removed in v0.41.

```

File: pennylane/workflow/qnode.py [hunk 2]
```diff
@@ -23,6 +23,7 @@
 from collections.abc import Callable, Sequence
 from typing import Any, Literal, Optional, Union, get_args
 
+import pytest
```
Comment (mudit2812) [inline pennylane/workflow/qnode.py:26]: Why are you using pytest 😅 . `warnings` is already imported.

And the rest of the code in this file needs to be changed accordingly as well.

File: tests/test_qnode.py [hunk 3]
```diff
@@ -177,11 +198,17 @@ def test_best_method_is_device(self):
 
         dev = CustomDeviceWithDiffMethod()
 
-        res = QNode.get_best_method(dev, "jax")
-        assert res == ("device", {}, dev)
+        with pytest.raises(
+            qml.PennyLaneDeprecationWarning, match="QNode.get_best_method is deprecated"
+        ):
```
Comment (mudit2812) [inline tests/test_qnode.py:203]: Same for the rest of the tests.

```suggestion
        with pytest.warns(
            qml.PennyLaneDeprecationWarning, match="QNode.get_best_method is deprecated"
        ):
```

**Review by mudit2812 — CHANGES_REQUESTED** [review]

File: pennylane/workflow/qnode.py (hunk 2 above)
Comment (andrijapau) [reply pennylane/workflow/qnode.py:26]: Sorry my brain is broken

File: pennylane/workflow/get_gradient_fn.py [hunk 4]
```diff
@@ -61,8 +63,10 @@ def _get_gradient_fn(
         )
 
     if diff_method == "best":
-        qn = qml.QNode(lambda: None, device, diff_method=None)
-        return qml.workflow.get_best_diff_method(qn)()
+        if tape and any(isinstance(o, qml.operation.CV) for o in tape):
+            return qml.gradients.param_shift_cv
+
+        return qml.gradients.param_shift
```
Comment (mudit2812) [inline pennylane/workflow/get_gradient_fn.py:69]: Why was this changed?

**Review by mudit2812 — COMMENTED** [review]
Just one more question, otherwise looks good.

**Comment by mudit2812** [comment]
Just remembered that `QNode.gradient_fn` is being removed in #6535, so your PR will have to adapt those changes as well.

File: pennylane/workflow/get_gradient_fn.py (hunk 4 above)
Comment (andrijapau) [reply pennylane/workflow/get_gradient_fn.py:69]: The `qml.workflow.get_best_diff_method` function was designed for user convenience, returning the gradient transform as a string. 

So I just took the logic from there and modified it to return the actual transform object instead of the string, aligning more closely with the purpose of `_get_gradient_fn`.

Comment (mudit2812) [reply pennylane/workflow/get_gradient_fn.py:69]: Thanks, resolving

**Review by mudit2812 — APPROVED** [review]
Looks good. I'm approving, but I suggest holding off on merging while #6535 is open, and syncing with @JerryChen97 to figure out the best merging strategy.

File: doc/development/deprecations.rst [hunk 5]
```diff
@@ -9,6 +9,12 @@ deprecations are listed below.
 Pending deprecations
 --------------------
 
+* The ``QNode.get_best_method``, ``QNode.best_method_str`` and ``QNode.get_gradient_fn`` methods have been deprecated. 
+  Instead, use the ``qml.workflow.get_best_diff_method`` and ``qml.workflow._get_gradient_fn`` functions.
```
Comment (albi3ro) [inline doc/development/deprecations.rst:13]:
```suggestion
  Instead, use the ``qml.workflow.get_best_diff_method``. 
```

Given we will can change the call signature or delete `_get_gradient_fn` with no notice, we shouldn't include it here.  I think `get_best_diff_method` would be sufficient for any user purposes.

If we end up locking down `resolve_execution_config` this cycle, we can update this deprecation to also point to that function, but I think for now `get_best_diff_method` is sufficient for public consumption.

Comment (andrijapau) [reply doc/development/deprecations.rst:13]: Good point!

File: doc/releases/changelog-dev.md [hunk 6]
```diff
@@ -50,6 +50,10 @@
 
 <h3>Deprecations 👋</h3>
 
+* The `QNode.get_best_method`, `QNode.best_method_str` and `QNode.get_gradient_fn` methods have been deprecated. 
+  Instead, use the `qml.workflow.get_best_diff_method` and `qml.workflow._get_gradient_fn` functions.
```
Comment (albi3ro) [inline doc/releases/changelog-dev.md:54]:
```suggestion
  Instead, use the `qml.workflow.get_best_diff_method` function.
```

**Review by albi3ro — APPROVED** [review]
❤️

File: tests/workflow/test_get_gradient_fn.py [hunk 7]
```diff
@@ -150,32 +150,39 @@ def test_finite_diff_method(self):
...
         gradient_fn = _get_gradient_fn(dev, diff_method="parameter-shift", tape=tape)
-        assert gradient_fn == qml.gradients.param_shift_cv
+        assert gradient_fn is qml.gradients.param_shift_cv
+
+    def test_best_method_with_cv_ops(self):
+        """Test that get_gradient_fn returns 'parameter-shift-cv' when CV operations are present on tape"""
+        dev = qml.device("default.gaussian", wires=1)
+        tape = qml.tape.QuantumScript([qml.Displacement(0.5, 0.0, wires=0)])
+        gradient_fn = _get_gradient_fn(dev, diff_method="best", tape=tape)
+        assert gradient_fn is qml.gradients.param_shift_cv
```
Comment (andrijapau) [inline tests/workflow/test_get_gradient_fn.py:185]: Added for coverage.

File: doc/development/deprecations.rst [hunk 8]
```diff
@@ -9,6 +9,12 @@ deprecations are listed below.
 Pending deprecations
 --------------------
 
+* The ``QNode.get_best_method``, ``QNode.best_method_str`` methods have been deprecated.
```
Comment (mudit2812) [inline doc/development/deprecations.rst:12]:
```suggestion
* The ``QNode.get_best_method`` and ``QNode.best_method_str`` methods have been deprecated. 
```

File: doc/releases/changelog-dev.md [hunk 9]
```diff
@@ -101,6 +101,10 @@
 
 <h3>Deprecations 👋</h3>
 
+* The `QNode.get_best_method`, `QNode.best_method_str` methods have been deprecated.
```
Comment (mudit2812) [inline doc/releases/changelog-dev.md:104]:
```suggestion
* The `QNode.get_best_method` and `QNode.best_method_str` methods have been deprecated. 
```

**Review by mudit2812 — APPROVED** [review]
